### ISSUE TITLE
Bug Report - SchemaError with apply Function in Polars 0.20.3

### ISSUE THREAD

**Issue by maazizzahra** [pola-rs/polars#14723]
### Checks

- [X] I have checked that this issue has not already been reported.
- [] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
lf = pl.LazyFrame({'a': ['[]']})
lf.select(pl.col('a').apply(eval, return_dtype=pl.List(pl.Utf8))).collect()
Polars + moderne :
lf.select(pl.col('a').map_elements(eval, return_dtype=pl.List(pl.String))).collect()
-> crash

I encountered the following error: polars.exceptions.SchemaError: expected output type 'List(String)', got 'List(Null)'; set return_dtype to the proper datatype
To resolve the issue, I used the following workaround:

lf.select(pl.when(pl.col('a') == '[]').then(pl.lit([]).cast(pl.List(pl.String))).otherwise(pl.col('a').map_elements(lambda x: eval(x) or [''], return_dtype=pl.List(pl.String)))).collect()

we use  Polars 0.20.3 
```

### Installed versions

<details>

```
I use 0.20.3```not 0.20.1

</details>

**Comment by MarcoGorelli** [comment]
thanks for the report, could you show an example of input data please

**Comment by maazizzahra** [comment]
lf = pl.LazyFrame({'a': ['[]']})
lf.select(pl.col('a').map_elements(eval, return_dtype=pl.List(pl.String))).collect()
we use  Polars 0.20.3

**Comment by cmdlineluser** [comment]
Is this the same issue that is discussed in https://github.com/pola-rs/polars/issues/10398#issuecomment-1672774735?

i.e. List(null) in this case should satisfy/upcast to List(String)

You can also force the dtype by manually wrapping in a Series inside the callback.

```python
lf.select(pl.col('a').map_elements(eval)).collect()
# shape: (1, 1)
# ┌────────────┐
# │ a          │
# │ ---        │
# │ list[null] │ # <- null
# ╞════════════╡
# │ []         │
# └────────────┘

lf.select(
    pl.col('a').map_elements(lambda x: pl.Series(eval(x), dtype=pl.String))
).collect()
# shape: (1, 1)
# ┌───────────┐
# │ a         │
# │ ---       │
# │ list[str] │ # <- str
# ╞═══════════╡
# │ []        │
# └───────────┘
```

**Comment by stinodego** [comment]
Should be fixed by #15699
